<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Цель-работы" data-toc-modified-id="Цель-работы-1">Цель работы</a></span></li><li><span><a href="#Исследование-таблиц." data-toc-modified-id="Исследование-таблиц.-2">Исследование таблиц.</a></span><ul class="toc-item"><li><span><a href="#books" data-toc-modified-id="books-2.1">books</a></span></li><li><span><a href="#authors" data-toc-modified-id="authors-2.2">authors</a></span></li><li><span><a href="#publishers" data-toc-modified-id="publishers-2.3">publishers</a></span></li><li><span><a href="#ratings" data-toc-modified-id="ratings-2.4">ratings</a></span></li><li><span><a href="#reviews" data-toc-modified-id="reviews-2.5">reviews</a></span></li></ul></li><li><span><a href="#Задания" data-toc-modified-id="Задания-3">Задания</a></span><ul class="toc-item"><li><span><a href="#Посчитайте,-сколько-книг-вышло-после-1-января-2000-года." data-toc-modified-id="Посчитайте,-сколько-книг-вышло-после-1-января-2000-года.-3.1">Посчитайте, сколько книг вышло после 1 января 2000 года.</a></span></li><li><span><a href="#Для-каждой-книги-посчитайте-количество-обзоров-и-среднюю-оценку." data-toc-modified-id="Для-каждой-книги-посчитайте-количество-обзоров-и-среднюю-оценку.-3.2">Для каждой книги посчитайте количество обзоров и среднюю оценку.</a></span></li><li><span><a href="#Определите-издательство,-которое-выпустило-наибольшее-число-книг-толще-50-страниц-—-так-вы-исключите-из-анализа-брошюры." data-toc-modified-id="Определите-издательство,-которое-выпустило-наибольшее-число-книг-толще-50-страниц-—-так-вы-исключите-из-анализа-брошюры.-3.3">Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры.</a></span></li><li><span><a href="#Определите-автора-с-самой-высокой-средней-оценкой-книг-—-учитывайте-только-книги-с-50-и-более-оценками." data-toc-modified-id="Определите-автора-с-самой-высокой-средней-оценкой-книг-—-учитывайте-только-книги-с-50-и-более-оценками.-3.4">Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками.</a></span></li><li><span><a href="#Посчитайте-среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-50-оценок." data-toc-modified-id="Посчитайте-среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-50-оценок.-3.5">Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.</a></span></li></ul></li></ul></div>

# Цель работы

Составить запросы SQL для данных, следуя заданиям.

# Исследование таблиц.

In [1]:
# импортируем библиотеки
import pandas as pd 
from sqlalchemy import create_engine 
# устанавливаем параметры
db_config = {'user': 'xxx', # имя пользователя
'pwd': 'xxx', # пароль
'host': 'xxx.net', 
'port': xxx, # порт подключения
'db': 'xxx'} # название базы данных

# Формируем строку соединения с БД.
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'], 
 db_config['pwd'], 
 db_config['host'], 
 db_config['port'], 
 db_config['db']) 
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

## books

In [2]:
# Формируем sql-запрос.
query = ''' SELECT *
            FROM books
            LIMIT 5;
        '''
pd.io.sql.read_sql(query, con = engine)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

**Таблица `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

## authors

In [3]:
query = ''' SELECT *
            FROM authors
            LIMIT 5;
        '''
pd.io.sql.read_sql(query, con = engine)

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

**Таблица `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

## publishers

In [4]:
query = ''' SELECT *
            FROM publishers
            LIMIT 5;
        '''
pd.io.sql.read_sql(query, con = engine)

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

## ratings

In [5]:
query = ''' SELECT *
            FROM ratings
            LIMIT 5;
        '''
pd.io.sql.read_sql(query, con = engine)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

## reviews

In [6]:
query = ''' SELECT *
            FROM reviews
            LIMIT 5;
        '''
pd.io.sql.read_sql(query, con = engine)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

# Задания

## Посчитайте, сколько книг вышло после 1 января 2000 года.

In [7]:
query = ''' SELECT COUNT(book_id)
            FROM books
            WHERE EXTRACT(YEAR FROM CAST(publication_date AS timestamp)) >= 2000;
        '''
pd.io.sql.read_sql(query, con = engine)

count
0    821

Из 1000 книг в данных  после 1 января 2000 года вышла 821 книга.

## Для каждой книги посчитайте количество обзоров и среднюю оценку.

In [8]:
query = ''' WITH
            rev AS (SELECT book_id,
                           COUNT(review_id) AS reviews    

                    FROM reviews
                    GROUP BY book_id),
            rat AS (SELECT book_id,
                           AVG(rating) AS avg_rating
                    FROM ratings
                    GROUP BY book_id)
                    
            SELECT b.book_id,
                   b.author_id,
                   b.title,
                   rev.reviews,
                   rat.avg_rating
            FROM books AS b
            LEFT JOIN rev ON b.book_id = rev.book_id
            LEFT JOIN rat ON rev.book_id = rat.book_id
            ORDER BY rev.reviews DESC
            LIMIT 10
                   
            
            
                        
        '''
pd.io.sql.read_sql(query, con = engine)

book_id  author_id                                              title  \
0      387        365                               Leonardo's Notebooks   
1      221        151                          Essential Tales and Poems   
2      672        144   The Cat in the Hat and Other Dr. Seuss Favorites   
3      191        568  Disney's Beauty and the Beast (A Little Golden...   
4       83        174    Anne Rice's The Vampire Lestat: A Graphic Novel   
5      808        350                            The Natural Way to Draw   
6      948        554                            Twilight (Twilight  #1)   
7      695        395  The Curious Incident of the Dog in the Night-Time   
8      673        235                             The Catcher in the Rye   
9      696        106             The Da Vinci Code (Robert Langdon  #2)   

   reviews  avg_rating  
0      NaN         NaN  
1      NaN         NaN  
2      NaN         NaN  
3      NaN         NaN  
4      NaN         NaN  
5      NaN         NaN  
6      7.0    3.662500  
7      6.0    4.081081  
8      6.0    3.825581  
9      6.0    3.830508

Неожиданно: "Сумерки" на первом месте по количеству обзоров - оценка правда всего 3.66.

## Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры.

In [9]:
query = ''' 
            WITH 
            book_filtered AS (SELECT *
                              FROM books
                              WHERE num_pages > 50)
            
            SELECT pub.publisher
            FROM (SELECT publisher,
                  COUNT(book_id) AS total_books
                  FROM book_filtered AS bf
                  LEFT JOIN publishers AS p ON bf.publisher_id = p.publisher_id
                  GROUP BY p.publisher
                  ORDER BY total_books DESC
                  LIMIT 1) AS pub;
                   
        '''
pd.io.sql.read_sql(query, con = engine)

publisher
0  Penguin Books

Penguin Books - Британское издательство основанное в 1935 году. C этим издательством стоит сотрудничать.

## Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками.

In [10]:
query = ''' WITH 
            br AS (SELECT book_id,
                          COUNT(rating) AS count,
                          AVG(rating) AS rating
                   FROM ratings
                   GROUP BY book_id
                   HAVING COUNT(rating)>50)
            SELECT a.author,
                   AVG(br.rating)
            FROM br
            LEFT JOIN books AS b ON br.book_id = b.book_id
            LEFT JOIN authors AS a ON b.author_id = a.author_id
            GROUP BY a.author_id
            ORDER BY AVG(br.rating) DESC
        '''
pd.io.sql.read_sql(query, con = engine)

author       avg
0                          J.K. Rowling/Mary GrandPré  4.283844
1                   Markus Zusak/Cao Xuân Việt Khương  4.264151
2                                      J.R.R. Tolkien  4.258446
3                                   Louisa May Alcott  4.192308
4                                        Rick Riordan  4.080645
5                                     William Golding  3.901408
6                                       J.D. Salinger  3.825581
7            Paulo Coelho/Alan R. Clarke/Özdemir İnce  3.789474
8   William Shakespeare/Paul Werstine/Barbara A. M...  3.787879
9                                           Dan Brown  3.754540
10                                         Lois Lowry  3.750000
11            George Orwell/Boris Grabnar/Peter Škerl  3.729730
12                                    Stephenie Meyer  3.662500
13                                     John Steinbeck  3.622951

Тут без сомнений - последние 20 лет только и слышно о Гарри Поттере.

## Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

In [11]:
query = ''' WITH
            users AS (SELECT username,
                             COUNT(rating_id) AS total_ratings
                      FROM ratings
                      GROUP BY username
                      HAVING COUNT(rating_id) > 50)
            SELECT AVG(final.total_reviews)
            FROM (SELECT u.username,
                         COUNT(r.review_id) AS total_reviews
                  FROM users AS u
                  LEFT JOIN reviews AS r ON u.username = r.username 
                  GROUP BY u.username) AS final;
            
            
        '''
pd.io.sql.read_sql(query, con = engine)

avg
0  24.333333

Пользователи, которые поставили больше 50 оценок, в среднем написали 24 обзора(и их всего 6).